Kernel/Graphics/VirtIOGPU/VirtIOGPUConsole.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2021, Sahan Fernando <sahan.h.fernando@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Graphics/VirtIOGPU/VirtIOGPUConsole.h>
+#include <Kernel/WorkQueue.h>
+
+namespace Kernel::Graphics {
+
+constexpr static AK::Time refresh_interval = AK::Time::from_milliseconds(16);
+
+void DirtyRect::union_rect(size_t x, size_t y, size_t width, size_t height)
+{
+    if (width == 0 || height == 0)
+        return;
+    if (m_is_dirty) {
+        m_x0 = min(x, m_x0);
+        m_y0 = min(y, m_y0);
+        m_x1 = max(x + width, m_x1);
+        m_y1 = max(y + height, m_y1);
+    } else {
+        m_is_dirty = true;
+        m_x0 = x;
+        m_y0 = y;
+        m_x1 = x + width;
+        m_y1 = y + height;
+    }
+}
+
+NonnullRefPtr<VirtIOGPUConsole> VirtIOGPUConsole::initialize(RefPtr<VirtIOGPU> gpu)
+{
+    return adopt_ref(*new VirtIOGPUConsole(gpu));
+}
+
+VirtIOGPUConsole::VirtIOGPUConsole(RefPtr<VirtIOGPU> gpu)
+    : GenericFramebufferConsole(gpu->framebuffer_width(), gpu->framebuffer_height(), gpu->framebuffer_pitch())
+    , m_gpu(gpu)
+{
+    m_framebuffer_region = gpu->framebuffer_region();
+    enqueue_refresh_timer();
+}
+
+void VirtIOGPUConsole::set_resolution(size_t width, size_t height, size_t)
+{
+    VERIFY(m_gpu->try_to_set_resolution(width, height));
```

Comment:
VERIFY statements with side effects are a bit weird. can we move it out of the verify?